### PR TITLE
MM-47074 : Migrate "components/suggestion/generic_channel_provider.jsx" to Typescript #21095

### PR DESCRIPTION
**Pull request by aashish0909**
=#### Summary

Migrate components/suggestion/generic_channel_provider.jsx to typescript

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/21095
Fixes https://mattermost.atlassian.net/browse/MM-47074

#### Related Pull Requests

NONE

#### Screenshots

NONE

#### Release Note

NONE
```release-note
NONE
```

### CODE REVIEW

**Comment by aashish0909** [comment]
I got a problem while migrating generic_channel_provider.jsx. 

There is an error in `components/interactive_dialog/dialog_element/dialog_element.tsx` file. I have attached the screenshot of the error below.

I couldn't figure out the solution for this. Please help me.

image

**Comment by aashish0909** [comment]
@sinansonmez @AshishDhama could you help me regarding this?

**Comment by hmhealey** [comment]
I think the error there is happening because `dialog_element.tsx` is declaring the type of `autocompleteChannels` incorrectly. The `error` callback should take a `ServerError` like in the one you added in `generic_channel_provider.tsx`, but the one defiled in the props for `DialogElement` does not take one.

That will probably be fixed by updating the type of `autocompleteChannels` in `DialogElement`'s `Props`. Hopefully that doesn't introduce any other type checking issues because typing Redux actions is very difficult with how we have that set up.

**Comment by aashish0909** [comment]
> I think the error there is happening because `dialog_element.tsx` is declaring the type of `autocompleteChannels` incorrectly. The `error` callback should take a `ServerError` like in the one you added in `generic_channel_provider.tsx`, but the one defiled in the props for `DialogElement` does not take one.
> 
> That will probably be fixed by updating the type of `autocompleteChannels` in `DialogElement`'s `Props`. Hopefully that doesn't introduce any other type checking issues because typing Redux actions is very difficult with how we have that set up.

Hey @hmhealey, I did what you said.
image


And it introduced a new error. :'(
image

File: components/interactive_dialog/dialog_element/dialog_element.tsx [hunk 1]
```diff
@@ -7,7 +7,7 @@ import {FormattedMessage} from 'react-intl';
```
Comment (sinansonmez) [inline components/interactive_dialog/dialog_element/dialog_element.tsx:7]: You need to change the type definition of `autocompleteChannels` 
- from `autocompleteChannels: (term: string, success: (channels: Channel[]) => void, error: () => void) => Promise<void>;`
- to `autocompleteChannels: (term: string, success: (channels: Channel[]) => void, error?: (err: ServerError) => void) => (ActionResult | Promise<ActionResult | ActionResult[]>);`

File: components/suggestion/generic_channel_provider.tsx [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
+export type Results = {
+    matchedPretext: string;
+    terms: string[];
+    items: Channel[];
+    component: React.ElementType;
+}
+
+type ResultsCallback = (results: Results) => void;
+
+type ChannelSearchFunc = (term: string, success: (channels: Channel[]) => void, error: (err: ServerError) => void) => ActionFunc
```
Comment (sinansonmez) [inline components/suggestion/generic_channel_provider.tsx:22]:
```suggestion
type ChannelSearchFunc = (term: string, success: (channels: Channel[]) => void, error?: (err: ServerError) => void) => (ActionResult | Promise<ActionResult | ActionResult[]>);
```

File: components/suggestion/generic_channel_provider.tsx [hunk 3]
```diff
@@ -47,28 +62,29 @@ class ChannelSuggestion extends Suggestion {
 }
 
 export default class ChannelProvider extends Provider {
-    constructor(channelSearchFunc) {
+    autocompleteChannels: any;
```
Comment (sinansonmez) [inline components/suggestion/generic_channel_provider.tsx:65]: as you already defined the type above, you can reuse it here
```suggestion
    autocompleteChannels: ChannelSearchFunc;
```

File: components/suggestion/generic_channel_provider.tsx [hunk 4]
```diff
@@ -47,28 +62,29 @@ class ChannelSuggestion extends Suggestion {
...
+    constructor(channelSearchFunc: (term: string, success: (channels: Channel[]) => void, error: (err: ServerError) => void) => ActionResult<any, any> | Promise<ActionResult<any, any> | Array<ActionResult<any, any>>>) {
```
Comment (sinansonmez) [inline components/suggestion/generic_channel_provider.tsx:66]: same as above
```suggestion
    constructor(channelSearchFunc: ChannelSearchFunc) {
```

**Review by sinansonmez — CHANGES_REQUESTED** [review]
You are almost there, as suggested by Harrison, after aligning type of `error` parameter, you need to align return type of `autocompleteChannels` function.

File: components/suggestion/generic_channel_provider.tsx (hunk 2 above)
Comment (aashish0909) [reply components/suggestion/generic_channel_provider.tsx:22]: Here, making `error` optional in ChannelSearchFunc gives an error in `action_menu.tsx`.

> Types of parameters 'error' and 'error' are incompatible.
Type '((err: ServerError) => void) | undefined' is not assignable to type '(err: ServerError) => void'.
Type 'undefined' is not assignable to type '(err: ServerError) => void'.

And if I don't make `error` optional to align with `action_menu.tsx`, i.e.
```suggestion
type ChannelSearchFunc = ((term: string, success: (channels: Channel[]) => void, error: (err: ServerError) => void) => (ActionResult | Promise<ActionResult | ActionResult[]>);
```

This error shows up.


image

Comment (sinansonmez) [reply components/suggestion/generic_channel_provider.tsx:22]: >Types of parameters 'error' and 'error' are incompatible.
Type '((err: ServerError) => void) | undefined' is not assignable to type '(err: ServerError) => void'.
Type 'undefined' is not assignable to type '(err: ServerError) => void'.

The reason you are getting the error is that you make it `optional` in one place and not in other places. Therefore you need to align error definition everywhere: `dialog_element`, `action_menu`, `generic_channel_provider` and `actions/channel_actions`. From the error message, try to understand where you are missing `error` definition

Comment (aashish0909) [reply components/suggestion/generic_channel_provider.tsx:22]: Oh. Got it👍🏻

**Comment by hmhealey** [comment]
This has been replaced by https://github.com/mattermost/mattermost-webapp/pull/12090 for anyone following it

**Pull request closed without merging**